### PR TITLE
Reorganize relocation table

### PR DESCRIPTION
**Pull request by kito-cheng**
Changes:
- Merge `Field` and `Calculation` to same column but different row.
- Update description from binutils's comment.

### CODE REVIEW

**Comment by kito-cheng** [comment]
@cmuellner

**Comment by kito-cheng** [comment]
Preview PDF fort this PR.
[riscv-abi.pdf](https://github.com/riscv-non-isa/riscv-elf-psabi-doc/files/7453952/riscv-abi.pdf)

**Comment by kito-cheng** [comment]
I am trying to squeeze more space from the relocation table, since @cmuellner suggest us could add a field to describe the type of relocation, static relocation or dynamic relocation, I think it's good idea too, but...the table already very compact...

**Comment by jrtc27** [comment]
Description and details serve basically the same purpose, merging them into a single description should help things. Using a fixed-width font for the relocation names might also help as it's smaller than the normal font.

**Comment by kito-cheng** [comment]
Changes:
- Merge `Description` and `Details` field.
- Use smaller size for `ELF Reloc Type`

Preview PDF:
[riscv-abi.pdf](https://github.com/riscv-non-isa/riscv-elf-psabi-doc/files/7459680/riscv-abi.pdf)

**Comment by cmuellner** [comment]
I agree, that `Description` and `Details` were not clearly distinguishable.
However, I think the intend of `Description` was a classification of the relocation and `Details` referred to (optional) additional details. The issue with the approach: the classification did not make sense because it was not related to anything (e.g. a subsection that defined the classes.

What still feels a bit odd: the table is in the toplevel part of the subsection 5.8. The definition of the contents of the table is in subsections thereof (5.8.1, 5.8.2, etc.). Typically this is done vice-versa: first introduce/define the terms and definitions and then use them.

My proposal:
* keep the classification of the relocations (runtime relocation, runtime TLS relocation, jump and call relocations, etc)
* 5.8: define what relocations are (what is this section about, where to find further information)
* 5.8.1: runtime relocations (describe their purpose and how they work; list all of them)
* 5.8.2: runtime TLS relocations
* ...
* 5.8.n: overview table of all relocations (no explanation needed; bonus points if the entries link to the classification description subsection above)

A similar approach can be found in the unpriv spec, where the chapter `RV32/64G Instruction Set Listings` provides an overview of all instructions after going through the details.

**Comment by kito-cheng** [comment]
Changes:
- Change `Description` to `Type`, and use `static` or `runtime` for the `Type` field.
- Move most `Description` to `Detail`

Preview PDF:
[riscv-abi.pdf](https://github.com/riscv-non-isa/riscv-elf-psabi-doc/files/7532231/riscv-abi.pdf)

**Comment by kito-cheng** [comment]
Changes:
- Add description for each column.
- Rename `Details` to `Description`.

**Comment by kito-cheng** [comment]
Preview PDF:
[riscv-abi.pdf](https://github.com/riscv-non-isa/riscv-elf-psabi-doc/files/7532356/riscv-abi.pdf)

File: riscv-elf.adoc [hunk 1]
```diff
@@ -184,68 +184,142 @@ sequences with their associated relocations in addition to the general purpose
...
 Global Offset Table or DWARF meta data.
 
-The following table provides details of the RISC-V ELF relocations (instruction
-specific relocations show the instruction type in the Details column):
+<<reloc-table>> provides details of the RISC-V ELF relocations, means of each
+columns are list below:
 
-[cols=">1,5,2,2,2,3"]
+Enum:: The enum value of the relocation, used in the r_info field of
+the ELF32_R_TYPE/ELF64_R_TYPE.
```
Comment (jrtc27) [inline riscv-elf.adoc:191]:
```suggestion
Enum:: The number of the relocation, encoded in the r_info field
```

File: riscv-elf.adoc [hunk 2]
```diff
@@ -184,68 +184,142 @@ sequences with their associated relocations in addition to the general purpose
 machine word sized relocations that are used for symbol addresses in the
 Global Offset Table or DWARF meta data.
 
-The following table provides details of the RISC-V ELF relocations (instruction
-specific relocations show the instruction type in the Details column):
+<<reloc-table>> provides details of the RISC-V ELF relocations, means of each
+columns are list below:
```
Comment (jrtc27) [inline riscv-elf.adoc:188]:
```suggestion
<<reloc-table>> provides details of the RISC-V ELF relocations; the meaning of each
column is given below:
```

File: riscv-elf.adoc [hunk 3]
```diff
@@ -184,68 +184,142 @@ sequences with their associated relocations in addition to the general purpose
...
 
-[cols=">1,5,2,2,2,3"]
+Enum:: The enum value of the relocation, used in the r_info field of
+the ELF32_R_TYPE/ELF64_R_TYPE.
+
+ELF Reloc Type:: The mnemonic name for the relocation.
```
Comment (jrtc27) [inline riscv-elf.adoc:193]:
```suggestion
ELF Reloc Type:: The name of the relocation
```

File: riscv-elf.adoc [hunk 4]
```diff
@@ -184,68 +184,142 @@ sequences with their associated relocations in addition to the general purpose
...
+
+Type:: Indicate this relocation is static or runtime relocation type:
```
Comment (jrtc27) [inline riscv-elf.adoc:195]:
```suggestion
Type:: Whether the relocation is a static or runtime relocation:
```

File: riscv-elf.adoc [hunk 5]
```diff
@@ -184,68 +184,142 @@ sequences with their associated relocations in addition to the general purpose
...
++
+- Static relocations are must resolved by static linker.
```
Comment (jrtc27) [inline riscv-elf.adoc:197]:
```suggestion
- Static relocations are always resolved by the static linker
```

File: riscv-elf.adoc [hunk 6]
```diff
@@ -184,68 +184,142 @@ sequences with their associated relocations in addition to the general purpose
...
+- Runtime relocations are resolved by static linker or dynamic linker.
```
Comment (jrtc27) [inline riscv-elf.adoc:198]:
```suggestion
- Runtime relocations can be resolved by both static and dynamic linkers
```

File: riscv-elf.adoc [hunk 7]
```diff
@@ -184,68 +184,142 @@ sequences with their associated relocations in addition to the general purpose
...
++
```
Comment (jrtc27) [inline riscv-elf.adoc:199]:
```suggestion
```

File: riscv-elf.adoc [hunk 8]
```diff
@@ -184,68 +184,142 @@ sequences with their associated relocations in addition to the general purpose
...
+
+Field::  Describe this relocation is apply to which field of data or
+         instruction, more detail see <<Field Symbols>>.
+
+Caculation::  Formula for how to resolve the relocation value, definition of
+              symbol can be found <<Calculation Symbols>>
```
Comment (jrtc27) [inline riscv-elf.adoc:205]:
```suggestion
Caculation:: Formula for how to resolve the relocation value; definitions of the
             symbols can be found in <<Calculation Symbols>>
```

File: riscv-elf.adoc [hunk 9]
```diff
@@ -184,68 +184,142 @@ sequences with their associated relocations in addition to the general purpose
...
+
+Description::  Description of this relocation.
```
Comment (jrtc27) [inline riscv-elf.adoc:207]:
```suggestion
Description:: Additional information about the relocation
```

File: riscv-elf.adoc [hunk 10]
```diff
@@ -184,68 +184,142 @@ sequences with their associated relocations in addition to the general purpose
...
+ELF Reloc Type:: The mnemonic name for the relocation.
+
+Type:: Indicate this relocation is static or runtime relocation type:
++
+- Static relocations are must resolved by static linker.
+- Runtime relocations are resolved by static linker or dynamic linker.
++
+
+Field::  Describe this relocation is apply to which field of data or
+         instruction, more detail see <<Field Symbols>>.
```
Comment (jrtc27) [inline riscv-elf.adoc:202]:
```suggestion
Field:: Describes the set of bits affected by this relocation; see <<Field Symbols>> for the definitions of the individual types
```

File: riscv-elf.adoc [hunk 11]
```diff
@@ -184,68 +184,142 @@ sequences with their associated relocations in addition to the general purpose
...
+                                                                  <|
+.2+| 1       .2+| [.relocname]#R_RISCV_32#            .2+| Runtime | _word32_          .2+| 32-bit relocation
+                                                                  <| S + A
+.2+| 2       .2+| [.relocname]#R_RISCV_64#            .2+| Runtime | _word64_          .2+| 64-bit relocation
+                                                                  <| S + A
+.2+| 3       .2+| [.relocname]#R_RISCV_RELATIVE#      .2+| Runtime | _wordclass_       .2+| Relocation against a local symbol in a shared object
+                                                                  <| B + A
+.2+| 4       .2+| [.relocname]#R_RISCV_COPY#          .2+| Runtime |                   .2+| Must be in executable; not allowed in shared library
+                                                                  <|
+.2+| 5       .2+| [.relocname]#R_RISCV_JUMP_SLOT#     .2+| Runtime | _wordclass_       .2+| Handled by PLT unless `LD_BIND_NOW`
```
Comment (jrtc27) [inline riscv-elf.adoc:225]: LD_BIND_NOW is more complicated than that. I'd just say something like "Indicates the symbol associated with a PLT entry".

File: riscv-elf.adoc [hunk 12]
```diff
@@ -184,68 +184,142 @@ sequences with their associated relocations in addition to the general purpose
...
+                                                                  <| S + A
+.2+| 29      .2+| [.relocname]#R_RISCV_TPREL_HI20#    .2+| Static  | _U-Type_          .2+| High 20 bits of TLS LE thread pointer offset, `%tprel_hi(symbol)`
+                                                                  <|
+.2+| 30      .2+| [.relocname]#R_RISCV_TPREL_LO12_I#  .2+| Static  | _I-Type_          .2+| Low 12 bits of TLS LE thread pointer offset, `%tprel_lo(symbol)`
+                                                                  <|
+.2+| 31      .2+| [.relocname]#R_RISCV_TPREL_LO12_S#  .2+| Static  | _S-Type_          .2+| Low 12 bits of TLS LE thread pointer offset, `%tprel_lo(symbol)`
+                                                                  <|
+.2+| 32      .2+| [.relocname]#R_RISCV_TPREL_ADD#     .2+| Static  |                   .2+| TLS LE thread pointer usage, `%tprel_add(symbol)`
+                                                                  <|
+.2+| 33      .2+| [.relocname]#R_RISCV_ADD8#          .2+| Static  | _word8_           .2+| 8-bit label addition .
```
Comment (jrtc27) [inline riscv-elf.adoc:273]: Stray . at the end?

**Review by jrtc27 — COMMENTED** [review]
Table formatting still isn't great but it's better than it was. I wonder if we should just drop the R_RISCV_ prefix from all the relocation names...

**Comment by kito-cheng** [comment]
Changes:
- Address @jrtc27's comment
- Drop `R_RISCV_` prefix in the relocation table.

Preview PDF:
[riscv-abi.pdf](https://github.com/riscv-non-isa/riscv-elf-psabi-doc/files/7593669/riscv-abi.pdf)

File: riscv-elf.adoc [hunk 13]
```diff
@@ -194,68 +194,139 @@ sequences with their associated relocations in addition to the general purpose
...
+
+Type:: Whether the relocation is a static or runtime relocation:
++
+- Static relocations are always resolved by the static linker
+- Runtime relocations can be resolved by both static and dynamic linkers
+
+Field:: Describes the set of bits affected by this relocation; see <<Field Symbols>> for the definitions of the individual types
+
+Caculation:: Formula for how to resolve the relocation value; definitions of the
+             symbols can be found in <<Calculation Symbols>>
```
Comment (jrtc27) [inline riscv-elf.adoc:212]:
```suggestion
Calculation:: Formula for how to resolve the relocation value; definitions of the
              symbols can be found in <<Calculation Symbols>>
```

**Review by jrtc27 — COMMENTED** [review]
Table looks about as readable as it's ever going to be when forced to fit in such a narrow space. Just one typo spotted then ready to squash and merge.

**Comment by kito-cheng** [comment]
Let me squash those commit after meeting :p

**Comment by kito-cheng** [comment]
Squashed

**Comment by jrtc27** [comment]
The Calculation typo fix got lost, and you've misspelled relocation in the new commit message

**Comment by kito-cheng** [comment]
Ooops, updated.

**Review by jrtc27 — APPROVED** [review]

**Pull request merged**